### PR TITLE
WIP: CMP implementation, incremental PR chunk 8: CMP generic and mock server

### PR DESCRIPTION
**Pull request by Akretsch**
Certificate Management Protocol (CMP, RFC 4210) extension to OpenSSL
Also includes CRMF (RFC 4211) and HTTP transfer (RFC 6712)

CMP and CRMF API is added to libcrypto, and the "cmp" app to the openssl CLI.
Adds extensive man pages and tests. Integration into build scripts.

8th chunk: CMP generic server and mock server for testing client in `crypto/cmp/cmp_server.c` and related files.
The testing client and related tests will be contributed in chunk 9. 

##### Checklist
- [x] documentation is added or updated

### CODE REVIEW

**Comment by Akretsch** [comment]
@mattcaswell We prepared a preview PR for CMP chunk 8. It shows the deltas to a frozen state of https://github.com/openssl/openssl/pull/10620

**Comment by Akretsch** [comment]
@mattcaswell Could  you please have a first view at this preliminary PR?

File: crypto/cmp/cmp_server.c [hunk 1]
```diff
@@ -0,0 +1,656 @@
...
+        }
+
+        /* check cert hash by recalculating it in place */
+        tmp = status->certHash;
+        status->certHash = NULL;
+        if (ossl_cmp_certstatus_set_certHash(status, srv_ctx->certOut))
+            res = status->certHash == NULL ? 0 /* avoiding SCA false positive */
+                : ASN1_OCTET_STRING_cmp(tmp, status->certHash) == 0;
+        ASN1_OCTET_STRING_free(status->certHash);
+        status->certHash = tmp;
```
Comment (mattcaswell) [inline crypto/cmp/cmp_server.c]: This seems fairly horrible. We seem to re-using the certHash value in status as a temporary place to store a value, just to resurrect it afterwards again. Is there no way to get the certHash without setting it in the status first?

File: crypto/cmp/cmp_server.c [hunk 2]
```diff
@@ -0,0 +1,656 @@
...
+    if ((*rsp = process_cb(srv_ctx, req)) == NULL)
+        return 0;
+
+    return 1;
+}
+
+/*
+ * Mocks the server connection. Works similar to OSSL_CMP_MSG_http_perform.
+ * A OSSL_CMP_SRV_CTX must be set as transfer_cb_arg
+ * returns 0 on success and else a CMP error reason code defined in cmp.h
```
Comment (mattcaswell) [inline crypto/cmp/cmp_server.c]: Hmmm! This is contrary to standard OpenSSL practice where 1 is success and 0 is an error. Why return a reason code and not just put an error on the stack?

**Comment by DDvO** [comment]
> I am struggling to understand why this code is in libcrypto. Is having a mock server considered a generally useful thing to do? I can see why we might want it in the apps - but why would other applications need this? To me this code seems better placed in the apps directory.

A really good and fundamental question here. We recently discussed it internally, too.
The idea was - and still is - that although currently the functionality is sufficient only for a test/mock server, it could be improved and extended towards productive code (building blocks) of a real server.

One may wonder if it could make sense in terms of development complexity to implement a full-fledged CMP server in C (or C++) on top of OpenSSL, and it turns out that some folks are actually considering this. If this wasn't the case I fully agree that all this code should be moved out of `crypto/`.

One could try separating code that clearly would make sense for a real server from code that is just good for the mock server, but given the rather small code base it would likely make more trouble (e.g., of defining a stable CMP server sub-API with callbacks for each type of requests etc.) than having all that in one file, `cmp_server.c`.

We might move the whole file to `apps/` and its header decls from `cmp.h` to `apps/include/` 
but then we could not use it for unit tests in `test/`.
We might instead move them to `test/` but then we could not use it for CLI-based tests.

File: crypto/cmp/cmp_server.c (hunk 1 above)
Comment (DDvO) [reply crypto/cmp/cmp_server.c]: I agree this is ugly. 
We'll improve that, which will likely require refactoring `ossl_cmp_certstatus_set_certHash()`.

File: crypto/cmp/cmp_server.c (hunk 2 above)
Comment (DDvO) [reply crypto/cmp/cmp_server.c]: I cannot recall what the design decision was, nor even who did it (@mpeylo or me),
and I see no strong reason for it except maybe simplicity.
We can change that (also in `cmp_client.c`, which be the core of chunk 9).

File: crypto/cmp/cmp_server.c (hunk 1 above)
Comment (Akretsch) [reply crypto/cmp/cmp_server.c]: see #206

File: crypto/cmp/cmp_server.c (hunk 2 above)
Comment (Akretsch) [reply crypto/cmp/cmp_server.c]: Fixed.

Comment (DDvO) [reply crypto/cmp/cmp_server.c]: @Akretsch, please just correct the comment as I indicated for the commit.

Comment (DDvO) [reply crypto/cmp/cmp_server.c]: @Akretsch, also the documentation needs to be update accordingly.

Comment (DDvO) [reply crypto/cmp/cmp_server.c]: ... but for now this pertains only to our internal branch, `cmp-dev`.

**Comment by mattcaswell** [comment]
IMO a mock server does not belong in libcrypto. If at some point in the future it is fleshed out to be more generally applicable then we can move it at that point.

> We might move the whole file to apps/ and its header decls from cmp.h to apps/include/
but then we could not directly use it for unit tests in test/.

I think it should be moved to the apps directory. There's already various instance of code in the apps directory which is used by the test code (for example opt.c in apps/lib is also used by the test code).

**Comment by DDvO** [comment]
@mattcaswell, all right, @Akretsch will move the server code to a `apps/cmp_server.c`.
This may imply the need to export a couple of further CMP functions - let's see.
We'll also look for a way to make use of that in both our API-level unit tests and CLI-based tests.

**Comment by DDvO** [comment]
I the meantime I'll also do a review of the mock server code.

File: crypto/cmp/build.info [hunk 3]
```diff
@@ -1,3 +1,4 @@
 LIBS=../../libcrypto
 SOURCE[../../libcrypto]= cmp_asn.c cmp_ctx.c cmp_err.c cmp_util.c \
-        cmp_status.c cmp_hdr.c cmp_protect.c cmp_msg.c cmp_vfy.c
+        cmp_status.c cmp_hdr.c cmp_protect.c cmp_msg.c cmp_vfy.c \
+        cmp_server.c
```
[the remark this answers is not in the document]
Comment (Akretsch) [reply crypto/cmp/build.info]: The  CMP (mock) server code is now in apps/

File: doc/man3/OSSL_CMP_SRV_CTX_new.pod [hunk 4]
```diff
@@ -0,0 +1,130 @@
+=pod
```
[the remark this answers is not in the document]
Comment (Akretsch) [reply doc/man3/OSSL_CMP_SRV_CTX_new.pod]: Moved.

File: crypto/cmp/cmp_server.c [hunk 5]
```diff
@@ -0,0 +1,658 @@
...
+
+    return msg;
+}
+
+static OSSL_CMP_MSG *process_error(OSSL_CMP_SRV_CTX *srv_ctx,
+                                   OSSL_CMP_MSG *req)
+{
+    OSSL_CMP_MSG *msg = ossl_cmp_pkiconf_new(srv_ctx->ctx);
+
+    (void)req; /* TODO make use of parameter */
```
[the remark this answers is not in the document]
Comment (Akretsch) [reply crypto/cmp/cmp_server.c]: Outdated

**Comment by Akretsch** [comment]
@mattcaswell We introduced a CMP (mock) server. Is there something else we should do?

File: include/openssl/cmp.h [hunk 6]
```diff
@@ -356,6 +356,32 @@ int OSSL_CMP_validate_cert_path(OSSL_CMP_CTX *ctx,
                                 X509_STORE *trusted_store, X509 *cert);
 int OSSL_CMP_print_cert_verify_cb(int ok, X509_STORE_CTX *ctx);
 
+/* from cmp_server.c */
```
[the remark this answers is not in the document]
Comment (DDvO) [reply include/openssl/cmp.h]: I moved part of the server related decls to `apps/cmp_mock_srv.h`,
renaming the mock server functions from `OSSL_CMP_SRV_CTX_*` to `ossl_cmp_mock_srv_*`.

File: crypto/cmp/cmp_server.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (DDvO) [reply crypto/cmp/cmp_server.c]: @Akretsch, this TODO has moved into `process_error()` of `apps/cmp_mock_src.c`.

File: doc/man3/OSSL_CMP_SRV_CTX_new.pod (hunk 4 above)
[the remark this answers is not in the document]
Comment (DDvO) [reply doc/man3/OSSL_CMP_SRV_CTX_new.pod]: The generic functions have stayed in `doc/man3/OSSL_CMP_SRV_CTX_new.pod` while
the mock server portion is now in `doc/internal/man3/ossl_cmp_mock_srv_new.pod`.

File: crypto/cmp/cmp_server.c [hunk 7]
```diff
@@ -0,0 +1,658 @@
+/*
```
[the remark this answers is not in the document]
Comment (DDvO) [reply crypto/cmp/cmp_server.c]: Only the mock server specific functions have move there.

File: include/openssl/cmp.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (DDvO) [reply include/openssl/cmp.h]: I've meanwhile moved that mock server header decls to `apps/cmp_mock_srv.h`
and change the function name prefixes to `ossl_cmp_mock_srv_`.

**Comment by DDvO** [comment]
Closing this preview PR since the actual PR is out now in https://github.com/openssl/openssl/pull/11142.
All comments received so far on it have been addressed. 
Thanks @mattcaswell!

**Pull request closed without merging**